viewing_party/party.py

```diff
@@ -1,23 +1,142 @@
+        return movie_dict
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+    
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data  
+
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+            return user_data
+    return user_data
+        
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    sum = 0.0
+    if not user_data["watched"]:
+        return sum
+    for movie in user_data["watched"]:
+        sum += movie["rating"]
+    average = sum / len(user_data["watched"])
+    return average
+
+def get_most_watched_genre(user_data):
+    genre_frequency_dict = {}
+    if not user_data["watched"]:
+        return None
+    for movie in user_data["watched"]:
+        if not movie["genre"] in genre_frequency_dict.keys():
+            genre_frequency_dict[movie["genre"]]= 1
+        else:
+            genre_frequency_dict[movie["genre"]] += 1
+    
+    most_watched_genre = max(genre_frequency_dict, key = genre_frequency_dict.get)
+    return most_watched_genre
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    user_watched_list = []
+    friend_watched_list = []
+    for movie in user_data["watched"]:
+        user_watched_list.append(movie["title"])
+
+    for friend in user_data["friends"]:
+        for friend_movie in friend["watched"]:
+            friend_watched_list.append(friend_movie["title"])
+
+    user_set = set(user_watched_list)
+    friend_set = set(friend_watched_list)
+    unique_set = user_set - friend_set
+
+    unique_list = []
+    for title in unique_set:
+        for movie in user_data["watched"]:
+            if title == movie["title"]:
+                unique_list.append(movie)
+    return unique_list
+
+def get_friends_unique_watched(user_data):
+    user_watched_list = []
+    friend_watched_list = []
+    for movie in user_data["watched"]:
+        user_watched_list.append(movie["title"])
+
+    for friend in user_data["friends"]:
+        for friend_movie in friend["watched"]:
+            friend_watched_list.append(friend_movie["title"])
+
+    user_set = set(user_watched_list)
+    friend_set = set(friend_watched_list)
+    unique_set = friend_set - user_set
+
+    unique_list = []
+    for title in unique_set:
+        for friend in user_data["friends"]:
+            for friend_movie in friend["watched"]:
+                if title == friend_movie["title"] and not (friend_movie in unique_list):
+                    unique_list.append(friend_movie)
+    return unique_list  

```

Comment:
Same kind of deal as before. This works and it's a great use of sets. However, it could definitely be pared down and refactored a bit to remove that triply nested loop on lines 98-102. There is this natural tendency to separate out a part of our dictionary to compare just the "titles" or just the "ratings, etc. More often than not, it's easier to just treat each "movie" dictionary as a whole item. You can absolutely check to see if a "movie" dictionary already exists in a list. With this in mind, feel free to compare entire dictionaries to each other! 

An algorithm using that particular strategy might include making a list of the unique movies in the friends lists. Then you can loop through the user_data["watched"] list and implement the same process as above!